### PR TITLE
CompanyService_ChangeCompany

### PR DESCRIPTION
**Pull request by KostinP**

### CODE REVIEW

File: DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+using LT.DigitalOffice.CompanyService.Models;
+using Microsoft.AspNetCore.Mvc;
+
+namespace LT.DigitalOffice.CompanyService.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class CompanyController : ControllerBase
+    {
+        [HttpPost("addCompany")]
```
Comment (lalagvana) [inline DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs]: Why "add"? Maybe "change"?

File: DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        public bool CreateUser([FromServices] IChangeCompanyCommand command, [FromBody] ChangeCompanyRequest request)
```
Comment (lalagvana) [inline DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs]: Why CreateUser?

File: DigitalOffice/Services/CompanyService/Commands/ChangeCompanyCommand.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            this.validator = validator;
+            this.repository = repository;
+            this.mapper = mapper;
+        }
+
+        public bool Execute(ChangeCompanyRequest request)
+        {
+            validator.ValidateAndThrow(request);
+
+            var user = mapper.Map(request);
```
Comment (MihailKulikov) [inline DigitalOffice/Services/CompanyService/Commands/ChangeCompanyCommand.cs]: user?

File: DigitalOffice/Services/CompanyService/Validators/ChangeCompanyValidator.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using FluentValidation;
+using LT.DigitalOffice.CompanyService.Models;
+
+namespace LT.DigitalOffice.CompanyService.Validators
+{
+    public class ChangeCompanyValidator : AbstractValidator<ChangeCompanyRequest>
+    {
+        public ChangeCompanyValidator()
+        {
+            RuleFor(company => company.CompanyId)
```
Comment (MihailKulikov) [inline DigitalOffice/Services/CompanyService/Validators/ChangeCompanyValidator.cs]: You should add rule for company.Name

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            validatorMock = new Mock<IValidator<ChangeCompanyRequest>>();
+            mapperMock = new Mock<IMapper<ChangeCompanyRequest, DbCompany>>();
+            repositoryMock = new Mock<ICompanyRepository>();
+
+            command = new ChangeCompanyCommand(validatorMock.Object, mapperMock.Object, repositoryMock.Object);
+        }
+
+
+        [Test]
+        public void FailCreateNewWorkTimeAccordingToValidator()
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Work time?

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            validatorMock
+                .Setup(x => x.Validate(It.IsAny<IValidationContext>()).IsValid)
+                .Returns(false);
+
+            Assert.Throws<ValidationException>(() => command.Execute(request));
+            repositoryMock.Verify(repository => repository.UpdateCompany(It.IsAny<DbCompany>()), Times.Never);
+        }
+
+        [Test]
+        public void FailCreateNewWorkTimeAccordingToRepository()
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Work time?

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+            repositoryMock
+                .Setup(x => x.UpdateCompany(It.IsAny<DbCompany>()))
+                .Throws(new Exception());
+
+            Assert.Throws<Exception>(() => command.Execute(request));
+        }
+
+        [Test]
+        public void SuccessfulCreateNewWorkTime()
```
Comment (MihailKulikov) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Work time?

**Review by MihailKulikov — CHANGES_REQUESTED** [review]

**Review by lalagvana — APPROVED** [review]

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs (hunk 7 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs (hunk 5 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs (hunk 6 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Validators/ChangeCompanyValidator.cs (hunk 4 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Validators/ChangeCompanyValidator.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Commands/ChangeCompanyCommand.cs (hunk 3 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Commands/ChangeCompanyCommand.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs (hunk 2 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs (hunk 1 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Controllers/CompanyController.cs]: Fixed.

**Comment by KostinP** [comment]
I didn't have time to fill in all the changes, the laptop sat down, eventually removed the mapper :(

File: DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                dbCompany.Name = company.Name;
+            }
+            if (company.IsActive != null)
+            {
+                var isActive = (bool)company.IsActive;
+                dbCompany.IsActive = isActive;
+            }
+            dbContext.SaveChanges();
+
+            return true;
```
Comment (mekk1t) [inline DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs]: Pavel, have you checked this in an actual Database?
I think this logic is broken. You just set object's properties, not the data in the actual database.

To update the DbEntity, you should use .Update(entity) method. Please, look it up. You'll get my point.

**Review by mekk1t — CHANGES_REQUESTED** [review]
UpdateCompany method just changes object's properties, not the actual data.

File: DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs (hunk 8 above)
Comment (mekk1t) [reply DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs]: Here's how I've tested it in my PR: 
image

**Comment by lalagvana** [comment]
Why is there no mapper?

File: DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs (hunk 8 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs]: Not relevant.

**Review by mekk1t — APPROVED** [review]

**Comment by KostinP** [comment]
changeCompany -> editCompany

File: DigitalOffice/Tests/CompanyServiceTests/Validators/ChangeCompanyValidatorTests.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+        [OneTimeSetUp]
+        public void OneTimeSetUp()
+        {
+            validator = new ChangeCompanyValidator();
+        }
+
+        [Test]
+        public void SuccessfulCompanyValidation()
+        {
```
Comment (djino97) [inline DigitalOffice/Tests/CompanyServiceTests/Validators/ChangeCompanyValidatorTests.cs]: New naming of tests

File: DigitalOffice/Tests/CompanyServiceTests/Repositories/CompanyRepositoryTests.cs [hunk 10]
```diff
@@ -96,10 +121,32 @@ public void ShouldRightGetCompanyById()
...
+        [Test]
+        public void ShouldUpdateCompany()
+        {
+            dbContext.Entry(dbCompanyInDb).State = EntityState.Detached;
+            var result = repository.UpdateCompany(dbCompanyToUpdate);
+
+            Assert.IsTrue(result);
+            var resultCompany = dbContext.Companies
+                .FirstOrDefaultAsync(x => x.Id == dbCompanyToUpdate.Id)
+                .Result;
```
Comment (djino97) [inline DigitalOffice/Tests/CompanyServiceTests/Repositories/CompanyRepositoryTests.cs]: First actions, then asserts

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            validatorMock = new Mock<IValidator<ChangeCompanyRequest>>();
+            mapperMock = new Mock<IMapper<ChangeCompanyRequest, DbCompany>>();
+            repositoryMock = new Mock<ICompanyRepository>();
+
+            command = new ChangeCompanyCommand(validatorMock.Object, mapperMock.Object, repositoryMock.Object);
+        }
+
+        [Test]
+        public void ShouldThrowExceptionAccordingToValidator()
+        {
```
Comment (djino97) [inline DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: ShouldThrowValidationExceptionWhenValidatorReturnsFalse (from your docs in the slack), where word `when`?

File: DigitalOffice/Services/CompanyService/Repositories/Interfaces/ICompanyRepository.cs [hunk 12]
```diff
@@ -23,6 +24,13 @@ public interface ICompanyRepository
         /// <returns>Id of the added company.</returns>
         Guid AddCompany(DbCompany company);
 
+        /// <summary>
+        /// Trying to update the company. Returns whether an update exited.
+        /// </summary>
+        /// <param name="company">Company to update.</param>
```
Comment (djino97) [inline DigitalOffice/Services/CompanyService/Repositories/Interfaces/ICompanyRepository.cs]: Edited company model

**Comment by djino97** [comment]
And why ChangeCompany, but not EditeCompany?

**Comment by KostinP** [comment]
> And why ChangeCompany, but not EditeCompany?

I haven't had time to fix it yet :)

File: DigitalOffice/Tests/CompanyServiceTests/Validators/ChangeCompanyValidatorTests.cs (hunk 9 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Validators/ChangeCompanyValidatorTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Repositories/CompanyRepositoryTests.cs (hunk 10 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Repositories/CompanyRepositoryTests.cs]: Fixed.

File: DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs (hunk 11 above)
Comment (KostinP) [reply DigitalOffice/Tests/CompanyServiceTests/Commands/ChangeCompanyCommandTests.cs]: Fixed.

File: DigitalOffice/Services/CompanyService/Repositories/Interfaces/ICompanyRepository.cs (hunk 12 above)
Comment (KostinP) [reply DigitalOffice/Services/CompanyService/Repositories/Interfaces/ICompanyRepository.cs]: Fixed.

**Review by ifelseelif — APPROVED** [review]

File: DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs [hunk 13]
```diff
@@ -18,42 +18,45 @@ public CompanyRepository(CompanyServiceDbContext dbContext)
...
-        public bool UpdateCompany(DbCompany company)
+        #region Company
+        public DbCompany GetCompanyById(Guid companyId)
         {
-            var dbCompany = dbContext.Companies
-                .AsNoTracking()
-                .FirstOrDefault(x => x.Id == company.Id);
-
+            var dbCompany = dbContext.Companies.FirstOrDefault(x => x.Id == companyId);
             if (dbCompany == null)
```
Comment (djino97) [inline DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs]: Add emty line

File: DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs [hunk 14]
```diff
@@ -66,12 +69,34 @@ public DbPosition GetPositionById(Guid positionId)
...
             dbContext.Positions.Add(newPosition);
             dbContext.SaveChanges();
 
             return newPosition.Id;
         }
+
+        public bool EditPosition(DbPosition newPosition)
+        {
+            var dbPosition = dbContext.Positions.FirstOrDefault(position => position.Id == newPosition.Id);
+            if (dbPosition == null)
```
Comment (djino97) [inline DigitalOffice/Services/CompanyService/Repositories/CompanyRepository.cs:88]: Add emty line

**Review by djino97 — APPROVED** [review]

**Review by MihailKulikov — APPROVED** [review]

**Pull request merged**